### PR TITLE
Fix Matrix(QRSparseQ) constructor

### LINKED ISSUE #505: Wrong full matrix for Q of QR-factorization in sparse
There seems to be a bug in the constructor `Matrix(::SuiteSparse.SPQR.QRSparseQ)`.

```
julia> versioninfo()
Julia Version 0.7.0-DEV.4473
Commit 2f826b55d6 (2018-03-06 05:55 UTC)
Platform Info:
  OS: Linux (x86_64-redhat-linux)
  CPU: Intel(R) Core(TM) i7-3610QM CPU @ 2.30GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-3.9.1 (ORCJIT, ivybridge)
Environment:

julia> A = sparse([0.0 1 0 0; 0 0 0 0])
2×4 SparseMatrixCSC{Float64,Int64} with 1 stored entry:
  [1, 2]  =  1.0

julia> F = qrfact(A);

julia> Matrix(F.Q)
2×0 Array{Float64,2}

julia> Matrix(F.Q')
2×2 Array{Float64,2}:
 1.0  0.0
 0.0  1.0
```
Size calculation of the result matrix in `LinearAlgebra/src/qr.jl:530` is not ok
```
Matrix(A::AbstractQ{T}) where {T} =
 lmul!(A, Matrix{T}(I, size(A.factors, 1), min(size(A.factors    )...)))
```
The number of columns of `Matrix(F.Q)` should be the same as the number of rows of `F.R`, or it should be square.

### LINKED ISSUE #26367: Wrong full matrix for Q of QR-factorization in sparse
There seems to be a bug in the constructor `Matrix(::SuiteSparse.SPQR.QRSparseQ)`.

```
julia> versioninfo()
Julia Version 0.7.0-DEV.4473
Commit 2f826b55d6 (2018-03-06 05:55 UTC)
Platform Info:
  OS: Linux (x86_64-redhat-linux)
  CPU: Intel(R) Core(TM) i7-3610QM CPU @ 2.30GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-3.9.1 (ORCJIT, ivybridge)
Environment:

julia> A = sparse([0.0 1 0 0; 0 0 0 0])
2×4 SparseMatrixCSC{Float64,Int64} with 1 stored entry:
  [1, 2]  =  1.0

julia> F = qrfact(A);

julia> Matrix(F.Q)
2×0 Array{Float64,2}

julia> Matrix(F.Q')
2×2 Array{Float64,2}:
 1.0  0.0
 0.0  1.0
```
Size calculation of the result matrix in `LinearAlgebra/src/qr.jl:530` is not ok
```
Matrix(A::AbstractQ{T}) where {T} =
 lmul!(A, Matrix{T}(I, size(A.factors, 1), min(size(A.factors    )...)))
```
The number of columns of `Matrix(F.Q)` should be the same as the number of rows of `F.R`, or it should be square.

### PR DESCRIPTION
**Pull request by andreasnoack**
Fixes JuliaLang/LinearAlgebra.jl#505

### CODE REVIEW

File: stdlib/LinearAlgebra/src/qr.jl [hunk 1]
```diff
@@ -527,21 +527,30 @@ AbstractMatrix{T}(Q::QRPackedQ) where {T} = QRPackedQ{T}(Q)
...
+    if 0 < dim
+        if dim <= 2
+            return size(getfield(Q, :factors), 1)
+        else
+            return 1
+        end
+    else
+        throw(BoundsError())
+    end
+end
```
Comment (stevengj) [inline stdlib/LinearAlgebra/src/qr.jl]: Seems easier to define this as:
```jl
size(Q::Union{QRCompactWYQ,QRPackedQ}, dim::Integer) = size(getfield(Q, :factors), dim == 2 ? 1 : dim)
```

File: stdlib/SuiteSparse/test/spqr.jl [hunk 2]
```diff
@@ -81,4 +81,9 @@ end
     @test A*xs ≈ A*xd
 end
 
+@testset "Issue 26367" begin
+    A = sparse([0.0 1 0 0; 0 0 0 0])
+    @test Matrix(qrfact(A).Q) == Matrix(I, 2, 2)
+end
```
Comment (stevengj) [inline stdlib/SuiteSparse/test/spqr.jl]: Also needs a test for the `QRCompactWYQ` case?

File: stdlib/LinearAlgebra/src/qr.jl (hunk 1 above)
Comment (KlausC) [reply stdlib/LinearAlgebra/src/qr.jl]: Are there cases possible, where Q is not a square? So why not
```
size(A::AbstractQ, dim::Integer) = 0 < dim ? (dim <= 2 ? size(getfield(A, :factors), 1) : 1) : throw(BoundsError())
```
was correct in my opinion.
The definition of `Matrix` could make use of `size` in this way:
```
Matrix(Q::AbstractQ{T}) where {T} = lmul!(Q, Matrix{T}(I, size(Q, 1), size(Q, 2)))
```
The changes in `SuiteSparse` are not needed in this case.

**Review by KlausC — CHANGES_REQUESTED** [review]
I propose to leave SuiteSparse unchanged and change the constructor of Matrix as noted in my comment.

File: stdlib/LinearAlgebra/src/qr.jl (hunk 1 above)
Comment (KlausC) [reply stdlib/LinearAlgebra/src/qr.jl]: Line 530 now produces a runtime error in the dense case. Also test case is missing to reveal that.

Comment (andreasnoack) [reply stdlib/LinearAlgebra/src/qr.jl]: It's unclear what "Line 530 now produces a runtime error in the dense case." means. If you have a specific example that fails then please post it.

Comment (andreasnoack) [reply stdlib/LinearAlgebra/src/qr.jl]: Regarding your first comment then Steve's version gives the same result as yours and mine. Steve's just reuses as much of `size(Matrix)` as possible which is nicer that my two if statements.

Comment (KlausC) [reply stdlib/LinearAlgebra/src/qr.jl]: The problem in line 530 is, that `T` in `Matrx{T}` is not defined:
```
julia> A = [1 2 1 1 1 1 ; 1 2 0 1 1 1; 1 2 0 1 1 1; 0 0 0 0 0 0];

julia> F = qrfact((Float64.(A)), Val(true))
QRPivoted{Float64,Array{Float64,2}}
Q factor:
4×4 LinearAlgebra.QRPackedQ{Float64,Array{Float64,2}}:
 -0.57735   0.816497  -8.75605e-17  0.0
 -0.57735  -0.408248  -0.707107     0.0
 -0.57735  -0.408248   0.707107     0.0
  0.0       0.0        0.0          1.0
R factor:
4×6 Array{Float64,2}:
 -3.4641  -0.57735   -1.73205  -1.73205  -1.73205  -1.73205
  0.0      0.816497   0.0       0.0       0.0       0.0    
  0.0      0.0        0.0       0.0       0.0       0.0    
  0.0      0.0        0.0       0.0       0.0       0.0    
permutation:
6-element Array{Int64,1}:
 2
 3
 1
 4
 5
 6

julia> Matrix(F.Q)
ERROR: UndefVarError: T not defined
Stacktrace:
 [1] Array{T,2} where T(::LinearAlgebra.QRPackedQ{Float64,Array{Float64,2}}) at /home/julia/julia/usr/share/julia/site/v0.7/LinearAlgebra/src/qr.jl:530
 [2] top-level scope

julia> VERSION
v"0.7.0-DEV.4516"

julia> versioninfo()
Julia Version 0.7.0-DEV.4516
Commit 1886be7a83 (2018-03-09 14:38 UTC)
Platform Info:
  OS: Linux (x86_64-redhat-linux)
  CPU: Intel(R) Core(TM) i7-3610QM CPU @ 2.30GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-3.9.1 (ORCJIT, ivybridge)
Environment:
```

**Comment by KlausC** [comment]
I just added some analysis to my post in discourse: [about QR-decomposition](https://discourse.julialang.org/t/about-qr-decomposition-linear-algebra/9606)

File: stdlib/LinearAlgebra/src/qr.jl (hunk 1 above)
Comment (andreasnoack) [reply stdlib/LinearAlgebra/src/qr.jl]: Thanks.

> Also test case is missing to reveal that.

The tests did reveal this though.

**Comment by KlausC** [comment]
> You can argue if that is a good idea but if one of the interpretations should go, it would definately be the Q0 interpretation.

I don't want to argue like that - when left multiplying with `F.Q*B` several row counts of B are acceptable. At the end, the rows of the rhs will have to be padded with zeroes, before applying the Householder factors.

Unfortunately I found another issue:
It is not possible to evaluate `R' * Q'`, which is the transpose of `Q*R`. I think right multiplication with `Q'` should also be defined for `AbstractQ` types.

> The confusing part is that size(Q, 2) != size(Matrix(Q), 2)

All I want here is clarity: that should be documented close to `qrfact`.

> We therefore have the choice of using either size(A,2) or rank(A) as the number of columns in Q0 and adjust R accordingly. 

I agree completely. In my post I suggest to make a decision about those case and also to store the row count of `R` in `Q`.

> In the dense case that is fairly simple because it is the same as the number of elementary reflectors, i.e. size(F.factors, 2) but in the sparse case that need not be the case since ...

I am not sure, if it is wise to rely on an implementation detail of the upstream library here. I would prefer to store the row count of  `R` in `Q` also in the LAPACK and generic cases.

> It estimates rank so when the trailing matrix is considered zero, no more reflectors are produced.

It would be easy, to add the rank-estimation also in the LAPACK and generic cases with pivoting: Just remove the last rows of `R`, if they are small. The corresponding Householder-factors would do no harm, because they operate on the complement of `image(A)`.

**Comment by andreasnoack** [comment]
> I am not sure, if it is wise to rely on an implementation detail of the upstream library here...
> ...It would be easy, to add the rank-estimation also in the LAPACK and generic cases with pivoting

In the dense case, the rank estimation is currently delayed until e.g. `ldiv!` which takes tolerance. This is how LAPACK is structured. I like the idea of keeping as much information as long as possible but it makes the sparse and dense cases behave slightly differently. The alternative is to move the rank estimation into the factorization in the dense case. This actually would be in line with pivoted Cholesky so actually LAPACK is a bit inconsistent here. I'll open separate issues for the insufficient documention that you pointed out and an issue for when the rank should be determined and then I'll merge this since it fixes a a current bug.

**Comment by andreasnoack** [comment]
Weird. I cannot reproduce the test errors locally.

File: stdlib/SuiteSparse/test/spqr.jl (hunk 2 above)
Comment (andreasnoack) [reply stdlib/SuiteSparse/test/spqr.jl]: That case should already be covered by the tests in LinearAlgebra

**Comment by JeffBezanson** [comment]
Is this really completely non-breaking?

**Comment by andreasnoack** [comment]
Yes. It shouldn't change any correct behavior.

**Comment by StefanKarpinski** [comment]
The policy for backports to long-term support branches is to be _extremely_ conservative. Can you elaborate on the details of what that means?

**Comment by andreasnoack** [comment]
It's a bugfix. Before this PR, we wrongly assumed that the number of reflectors was the same as the number of columns in a matrix. That isn't the case for sparse matrices so a field is added to the struct to indicate number of columns.

**Pull request merged**